nixos/modules/services/networking/mbim-network.nix

```
@@ -0,0 +1,126 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.mbim-network;
+
+  # Transforms "/dev/cdc-wdm0" into "mbim-network-dev-cdc-wdm0" for file and
+  # systemd unit names.
+  cleanName = name: "mbim-network${replaceStrings [ "/" ] [ "-" ] name}";
+
+  makeSystemdUnit = dcfg: name:
+    let
+      path = getBin pkgs.libmbim;
+
+      # Config file format defined in 'man 1 mbim-network'.
+      configFile = with dcfg;
+        pkgs.writeText "${cleanName name}.conf" ''
+          APN=${apn}
+          PROXY=${if cfg.proxy then "yes" else "no"}
+          ${optionalString (apnUser != "") "APN_USER=${apnUser}"}
+          ${optionalString (apnPass != "") "APN_PASS=${apnPass}"}
+          ${optionalString (apnAuth != "") "APN_AUTH=${apnAuth}"}
+        '';
+
+      # Mimic the format of the commands run by mbim-network, including the
+      # optional use of mbim-proxy.
+      cliCmd = "${path}/bin/mbimcli -d ${name}${optionalString cfg.proxy " --device-open-proxy"}";
+      netCmd = "${path}/bin/mbim-network --profile ${configFile} ${name}";
+    in {
+      description = "mbim-network for '${name}'";
+      wantedBy = [ "multi-user.target" ];
+      path = [ path ];
+
+      serviceConfig = {
+        # 'mbim-network start' is a one-shot command which establishes the proper
+        # state for MBIM modem communications. The equivalent 'stop' command
+        # tears down the connection when the systemd unit stopped.
+        User = "root";
+        Type = "oneshot";
+        RemainAfterExit = "yes";
+        ExecStart = "${netCmd} start";
+        ExecStop = "${netCmd} stop";
+      };
+
+      # Ensure the connection was started and an IP configuration can be fetched.
+      postStart = ''
+        ${netCmd} status | grep -q 'Status: activated'
+        ${cliCmd} --query-ip-configuration
+      '';
+      postStop = "${netCmd} status | grep -q 'Status: deactivated'";
+    };
+in {
+  options.services.mbim-network = {
+    # Allow configuring zero or more MBIM devices with differing configurations.
+    devices = mkOption {
+      default = { };
+      example = literalExample ''"/dev/cdc-wdm0".apn = "internet";'';
```

```suggestion
      example = { "/dev/cdc-wdm0".apn = "internet"; };
```
At least, the example should include the braces.

---

Done.